src/parachain/loans.ts

```diff
@@ -268,12 +291,48 @@ export class DefaultLoansAPI implements LoansAPI {
         );
     }
 
+    async getBorrowerAccountIds(): Promise<Array<AccountId>> {
+        const accountBorrows = await this.api.query.loans.accountBorrows.entries();
+        return [
+            // Even if two `AccountId`s store the same ID, the actual objects will not be equal when compared,
+            // so need to use the string representation
+            ...new Set(accountBorrows.map((key) => storageKeyToNthInner(key[0], 1).toString())),
+        ].map((accountIdString, _index, _arr) => newAccountId(this.api, accountIdString));
+    }
+
+    async getUndercollateralizedBorrowers(): Promise<Array<UndercollateralizedPosition>> {
+        const borrowers = await this.getBorrowerAccountIds();
+        const [liquidity, borrows, collateral] = await Promise.all([
+            Promise.all(borrowers.map(this.getLiquidationThresholdLiquidity.bind(this))),
+            Promise.all(borrowers.map(this.getBorrowPositionsOfAccount.bind(this))),
+            Promise.all(borrowers.map(this.getLendPositionsOfAccount.bind(this))),
+        ]);
+        const undercollateralizedPositions: Array<UndercollateralizedPosition> = [];
+        for (let i = 0; i < borrowers.length; i++) {
+            undercollateralizedPositions.push({
+                accountId: borrowers[i],
+                shortfall: liquidity[i].shortfall,
+                collateralPositions: collateral[i],
+                borrowPositions: borrows[i],
+            });
+        }
+        return undercollateralizedPositions.filter((position) => !position.shortfall.isZero());
+    }
+
+    async getLiquidationThresholdLiquidity(accountId: AccountId): Promise<AccountLiquidity> {
+        const [rawLiquidity, rawShortfall] = await this.api.rpc.loans.getLiquidationThresholdLiquidity(accountId);
```

Comment:
Does this always return a value for any account?

Reply:
Good point - yes. I added this getter to the interface so I can document it, also added some doc comments to the `AccountLiquidity` type.